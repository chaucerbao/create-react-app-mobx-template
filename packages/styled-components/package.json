{
  "name": "styled-components-styles",
  "version": "0.1.0",
  "description": "Using styled-components to manage the styles",
  "author": "Alvin Chan <no@email.com>",
  "license": "ISC",
  "stylelint": {
    "processors": [
      "stylelint-processor-styled-components"
    ],
    "extends": [
      "stylelint-config-standard",
      "stylelint-config-concentric-order",
      "stylelint-config-styled-components"
    ]
  },
  "devDependencies": {
    "source-map-explorer": "^1.5.0",
    "stylelint": "^9.2.0",
    "stylelint-config-concentric-order": "^3.1.0",
    "stylelint-config-standard": "^18.2.0",
    "stylelint-config-styled-components": "^0.1.1",
    "stylelint-processor-styled-components": "^1.3.1",
    "tslint": "^5.10.0",
    "webpack-cli": "^2.1.3",
    "webpack-dev-server": "^3.1.4"
  },
  "dependencies": {
    "styled-components": "^3.2.6"
  },
  "scripts": {
    "start": "webpack-dev-server --config ../../webpack.config.js -d",
    "build": "webpack-cli --config ../../webpack.config.js -p --devtool source-map --progress",
    "analyze": "source-map-explorer public/js/script.js{,.map}"
  }
}
