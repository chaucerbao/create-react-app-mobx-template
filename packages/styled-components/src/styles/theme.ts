// tslint:disable object-literal-sort-keys

// Type definitions
export interface ITheme {
  breakpoints: IDictionary
  colors: IDictionary
  spacers: IDictionary
}
interface IDictionary {
  [key: string]: number | string
}

// Theme
const breakpoints = {
  mobile: '0',
  tablet: '768px',
  desktop: '1280px',
}

const colors = {
  primary: '#444',
  grey: '#888',
  white: '#fafafa',
}

const spacers = {
  xs: '4px',
  sm: '8px',
  md: '16px',
  lg: '24px',
  xl: '32px',
}

const sort = (dictionary: IDictionary) =>
  Object.entries(dictionary)
    .sort((a, b) => {
      const aValue = parseFloat(a[1].toString())
      const bValue = parseFloat(b[1].toString())

      if (aValue < bValue) {
        return -1
      }
      if (aValue > bValue) {
        return 1
      }

      return 0
    })
    .reduce((sortedDictionary: IDictionary, definition) => {
      sortedDictionary[definition[0]] = definition[1]

      return sortedDictionary
    }, {})

const theme: ITheme = {
  breakpoints: sort(breakpoints),
  colors,
  spacers: sort(spacers),
}

// Exports
export default theme
