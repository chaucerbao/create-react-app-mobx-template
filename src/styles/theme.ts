// Styles
export default {
  black: '#000',
  blue: 'steelblue',
  darkblue: 'royalblue',
  darkgray: '#333',
  dimgray: 'dimgray',
  lightblue: 'lightsteelblue',
  lightgray: 'gainsboro',
  red: 'red',
  white: '#fff'
}
