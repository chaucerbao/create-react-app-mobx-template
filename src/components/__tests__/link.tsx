// Third-party dependencies
import { mount } from 'enzyme'
import { Provider } from 'mobx-react'
import * as React from 'react'
import * as renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'

// Local dependencies
import theme from '../../styles/theme'

// Component
import Link from '../link'

// Mocks
const stores = { viewStore: { goTo: jest.fn() } }

describe('<Link />', () => {
  it('renders an internal Link', () => {
    const component = renderer
      .create(
        <Provider stores={stores}>
          <ThemeProvider theme={theme}>
            <Link to="/url">Link Text</Link>
          </ThemeProvider>
        </Provider>
      )
      .toJSON()

    expect(component).toMatchSnapshot()
  })

  it('renders an external Link', () => {
    const component = renderer
      .create(
        <Provider stores={stores}>
          <ThemeProvider theme={theme}>
            <Link to="http://domain/">Link Text</Link>
          </ThemeProvider>
        </Provider>
      )
      .toJSON()

    expect(component).toMatchSnapshot()
  })

  it('renders an internal pop-up Link', () => {
    const component = renderer
      .create(
        <Provider stores={stores}>
          <ThemeProvider theme={theme}>
            <Link to="/url" target="_blank">
              Link Text
            </Link>
          </ThemeProvider>
        </Provider>
      )
      .toJSON()

    expect(component).toMatchSnapshot()
  })

  it('renders an external pop-up Link', () => {
    const component = renderer
      .create(
        <Provider stores={stores}>
          <ThemeProvider theme={theme}>
            <Link to="http://domain/" target="_blank">
              Link Text
            </Link>
          </ThemeProvider>
        </Provider>
      )
      .toJSON()

    expect(component).toMatchSnapshot()
    expect(component!.props.rel).toBe('noopener')
  })

  it('renders an external pop-up Link with `rel` attribute overridden', () => {
    const component = renderer
      .create(
        <Provider stores={stores}>
          <ThemeProvider theme={theme}>
            <Link to="http://domain/" target="_blank" rel="nofollow">
              Link Text
            </Link>
          </ThemeProvider>
        </Provider>
      )
      .toJSON()

    expect(component).toMatchSnapshot()
  })
})

describe('<Link />', () => {
  beforeEach(() => {
    stores.viewStore.goTo.mockReset()
  })

  it('runs an onClick handler, and navigates', () => {
    const onClickHandler = jest.fn()
    const component = mount(
      <Provider stores={stores}>
        <ThemeProvider theme={theme}>
          <Link to="/url" onClick={onClickHandler}>
            Link Text
          </Link>
        </ThemeProvider>
      </Provider>
    )

    expect(onClickHandler.mock.calls.length).toBe(0)
    expect(stores.viewStore.goTo.mock.calls.length).toBe(0)

    component.simulate('click')

    expect(onClickHandler.mock.calls.length).toBe(1)
    expect(stores.viewStore.goTo.mock.calls.length).toBe(1)
    expect(stores.viewStore.goTo.mock.calls[0][0]).toBe('/url')
  })

  it('runs an onClick handler that returns false, and does not navigate', () => {
    const onClickHandler = jest.fn().mockReturnValue(false)
    const component = mount(
      <Provider stores={stores}>
        <ThemeProvider theme={theme}>
          <Link to="/url" onClick={onClickHandler}>
            Link Text
          </Link>
        </ThemeProvider>
      </Provider>
    )

    expect(onClickHandler.mock.calls.length).toBe(0)
    expect(stores.viewStore.goTo.mock.calls.length).toBe(0)

    component.simulate('click')

    expect(onClickHandler.mock.calls.length).toBe(1)
    expect(stores.viewStore.goTo.mock.calls.length).toBe(0)
  })

  it('does not run the onClick handler, nor navigate, if disabled', () => {
    const onClickHandler = jest.fn()
    const component = mount(
      <Provider stores={stores}>
        <ThemeProvider theme={theme}>
          <Link to="/url" disabled={true}>
            Link Text
          </Link>
        </ThemeProvider>
      </Provider>
    )

    expect(onClickHandler.mock.calls.length).toBe(0)
    expect(stores.viewStore.goTo.mock.calls.length).toBe(0)

    component.simulate('click')

    expect(onClickHandler.mock.calls.length).toBe(0)
    expect(stores.viewStore.goTo.mock.calls.length).toBe(0)
  })
})
