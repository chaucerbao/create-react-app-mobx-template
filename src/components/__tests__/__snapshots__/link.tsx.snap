// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> renders an external Link 1`] = `
.c0 {
  text-decoration: none;
  color: steelblue;
}

.c0:hover {
  text-decoration: underline;
}

.c0:active {
  color: lightsteelblue;
}

<a
  className="c0"
  href="http://domain/"
  onClick={[Function]}
  rel={undefined}
>
  Link Text
</a>
`;

exports[`<Link /> renders an external pop-up Link 1`] = `
.c0 {
  text-decoration: none;
  color: steelblue;
}

.c0:hover {
  text-decoration: underline;
}

.c0:active {
  color: lightsteelblue;
}

<a
  className="c0"
  href="http://domain/"
  onClick={[Function]}
  rel="noopener"
  target="_blank"
>
  Link Text
</a>
`;

exports[`<Link /> renders an external pop-up Link with \`rel\` attribute overridden 1`] = `
.c0 {
  text-decoration: none;
  color: steelblue;
}

.c0:hover {
  text-decoration: underline;
}

.c0:active {
  color: lightsteelblue;
}

<a
  className="c0"
  href="http://domain/"
  onClick={[Function]}
  rel="nofollow"
  target="_blank"
>
  Link Text
</a>
`;

exports[`<Link /> renders an internal Link 1`] = `
.c0 {
  text-decoration: none;
  color: steelblue;
}

.c0:hover {
  text-decoration: underline;
}

.c0:active {
  color: lightsteelblue;
}

<a
  className="c0"
  href="/url"
  onClick={[Function]}
  rel={undefined}
>
  Link Text
</a>
`;

exports[`<Link /> renders an internal pop-up Link 1`] = `
.c0 {
  text-decoration: none;
  color: steelblue;
}

.c0:hover {
  text-decoration: underline;
}

.c0:active {
  color: lightsteelblue;
}

<a
  className="c0"
  href="/url"
  onClick={[Function]}
  rel={undefined}
  target="_blank"
>
  Link Text
</a>
`;
